src/core/iomgr/reachability.h

```diff
@@ -0,0 +1,114 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+#define GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+
+// Reachability (as used here) is defined as being the existance of a network
+// route to a specified address. It does not indicate whether a server is
+// serving traffic at that address.
+
+/// Network type: a broad classification of different kinds of network
+/// Useful for systems that want to send traffic only on a restricted class
+typedef enum grpc_network_type {
+  /// Network type is unknown
+  GRPC_NETWORK_UNKNOWN_TYPE,
+  /// Network is a wired connection
+  GRPC_NETWORK_WIRED,
+  /// Network is a wifi connection
+  GRPC_NETWORK_WIFI,
+  /// Network is a cellular connection
+  GRPC_NETWORK_CELLULAR
+} grpc_network_type;
+
+/// A description of a nic
+typedef struct grpc_network_interface {
+  /// Type of the network this nic refers to
+  grpc_network_type network_type;
+  /// Address to pass to bind() to select this nic
+  const sockaddr_t *bind_addr;
+  size_t bind_addr_len;
+  /// A name a sysadmin could use to identify this nic
+  const char *name;
+} grpc_network_interface;
+
+/// Copy a network interface description
+grpc_network_interface *grpc_network_interface_copy(
+    grpc_network_interface *src);
+/// Destroy a network interface description
+void grpc_network_interface_destroy(grpc_network_interface *nic);
+/// Enumerate all network interfaces on this system
+void grpc_enumerate_network_interfaces(grpc_network_interface **nics,
+                                       size_t *nic_count);
+
+/// One time global initialization
+void grpc_reachability_init(void);
+/// Global shutdown - cancels all outstanding watches
+void grpc_reachability_shutdown(void);
+
+/// Watch \a addr for reachability.
+/// If \a addr == NULL, watch for any network being available.
+/// When \a addr becomes reachable, call \a on_reachable with success=true
+/// \a optionally_only_on_nic allows reachability to be awaited only via a
+/// specific network interface
+void grpc_reachability_await_reachable(
+    const sockaddr_t *addr, size_t addr_len,
+    const grpc_network_interface *optionally_only_on_nic,
+    grpc_closure *on_reachable);
+/// Watch \a addr for reachability.
+/// If \a addr == NULL, watch for all networks being unavailable.
+/// When \a addr becomes unreachable, call \a on_unreachable with success=true
+/// \a optionally_only_on_nic allows reachability to be awaited only via a
+/// specific network interface
+void grpc_reachability_await_unreachable(
+    const sockaddr_t *addr, size_t addr_len,
+    const grpc_network_interface *optionally_only_on_nic,
+    grpc_closure *on_unreachable);
+
+/// Query (instantaneous) reachability of \a addr
```

Comment:
A detail on the resolution algorithm, for Apple's implementation:

In some cases (like the device being in low-power mode), `awaitReachable` might await indefinitely, even though as soon as a connection were attempted it would succeed (and `awaitReachable` trigger). So one should always attempt to connect when the application requires it, and then query reachability.